SU2_CFD/include/numerics/turbulent/turb_sources.hpp

```diff
@@ -498,58 +496,67 @@ class CCompressibilityCorrection final : public ParentClass {
- * \note From NASA Turbulence model site. http://turbmodels.larc.nasa.gov/spalart.html
- * This form was developed primarily to improve the near-wall numerical behavior of the model (i.e. the goal was to
- * improve the convergence behavior). The reference is: Edwards, J. R. and Chandra, S. "Comparison of Eddy
- * Viscosity-Transport Turbulence Models for Three-Dimensional, Shock-Separated Flowfields," AIAA Journal, Vol. 34, No.
- * 4, 1996, pp. 756-763.
- */
 template <class FlowIndices>
-using CSourcePieceWise_TurbSA_E = CSourceBase_TurbSA<FlowIndices, detail::Omega::Edw, detail::ft2::Zero,
-                                                     detail::ModVort::Edw, detail::r::Edw, detail::SourceTerms::Bsl>;
+struct EdwardsSA {
+  template <class Ft2>
+  using type = CSourceBase_TurbSA<FlowIndices, Omega::Edw, Ft2, ModVort::Edw, r::Edw, SourceTerms::Bsl>;
+};
 
-/*!
- * \class CSourcePieceWise_TurbSA_E_COMP
- * \brief Class for integrating the source terms of the Spalart-Allmaras model with Edwards modification
- *        and compressibility correction.
- */
-template <class FlowIndices>
-using CSourcePieceWise_TurbSA_E_COMP = detail::CCompressibilityCorrection<CSourcePieceWise_TurbSA_E<FlowIndices>>;
+template <class BaseType, class... Ts>
+CNumerics* AddCompressibilityCorrection(bool use_comp, Ts... args) {
+  if (use_comp) {
+    return new detail::CCompressibilityCorrection<BaseType>(args...);
+  } else {
+    return new BaseType(args...);
+  }
+}
+
+template <class BaseType, class... Ts>
+CNumerics* SAFactoryImpl(bool use_ft2, Ts... args) {
+  if (use_ft2) {
+    return AddCompressibilityCorrection<typename BaseType::template type<ft2::Nonzero>>(args...);
+  } else {
+    return AddCompressibilityCorrection<typename BaseType::template type<ft2::Zero>>(args...);
+  }
+}
+
+}  // namespace detail
 
 /*!
- * \class CSourcePieceWise_TurbSA_Neg
- * \brief Class for integrating the source terms of the negative Spalart-Allmaras model.
+ * \brief Creates an SA source based on the version and modifications/correction in the config.
  */
 template <class FlowIndices>
-using CSourcePieceWise_TurbSA_Neg = CSourceBase_TurbSA<FlowIndices, detail::Omega::Bsl, detail::ft2::Zero,
-                                                       detail::ModVort::Neg, detail::r::Bsl, detail::SourceTerms::Neg>;
+CNumerics* SAFactory(unsigned short nDim, const CConfig* config) {
+  const auto options = config->GetSAParsedOptions();
+
+  switch (options.version) {
+    case SA_OPTIONS::NONE:
+      return detail::SAFactoryImpl<detail::BaselineSA<FlowIndices>>(options.ft2, options.comp, nDim, config);
+    case SA_OPTIONS::NEG:
+      return detail::SAFactoryImpl<detail::NegativeSA<FlowIndices>>(options.ft2, options.comp, nDim, config);
+    case SA_OPTIONS::EDW:
+      return detail::SAFactoryImpl<detail::EdwardsSA<FlowIndices>>(options.ft2, options.comp, nDim, config);
```

Comment:
3 flow index types * 3 base versions * 2 compressible or not * 2 ft2 or not = 36 types of SA source being compiled xD

Reply:
I think we need more.